components/plurals/src/rules/runtime/ast.rs

```diff
@@ -250,89 +265,50 @@ impl fmt::Display for Rule<'_> {
-///  ▲ ▲
-///  │ │
-///  │ Polarity
-///  │
-///  AndOr
-/// ```
-///
-/// # Optimization
-///
-/// This model is optimized for efficient packaging of the `Relation` elements
-/// and performant deserialization from the `RelationULE` to `Relation` type.
-///
-/// # Constraints
-///
-/// The model constraints the `Operand` to 64 variants, and `Modulo` to `u32::MAX`.
-#[derive(Debug, PartialEq)]
-#[repr(packed)]
-pub struct RelationULE {
-    /// This maps to (AndOr, Polarity, Operand),
-    /// with the first bit mapping to AndOr (1 == And), the second bit
-    /// to Polarity (1 == Positive), and the remaining bits to Operand
-    /// encoded via Operand::encode. It is unsound for the Operand bits to
-    /// not be a valid encoded Operand.
-    andor_polarity_operand: u8,
-    modulo: <u32 as AsULE>::ULE,
-    range_list: [RangeOrValueULE],
-}
-
 impl RelationULE {
     #[inline]
     pub fn as_relation(&self) -> Relation {
-        let (and_or, polarity, operand) =
-            unsafe { Self::decode_andor_polarity_operand(self.andor_polarity_operand) };
-        Relation {
-            and_or,
-            polarity,
-            operand,
-            modulo: u32::from_unaligned(self.modulo),
-            range_list: ZeroVec::Borrowed(&self.range_list),
-        }
+        zerofrom::ZeroFrom::zero_from(self)
     }
+}
 
-    fn encode_andor_polarity_operand(and_or: AndOr, polarity: Polarity, operand: Operand) -> u8 {
-        let encoded_operand = u8::from(operand);
-        debug_assert!(encoded_operand <= 0b0011_1111);
-        (((and_or == AndOr::And) as u8) << 7)
-            + (((polarity == Polarity::Positive) as u8) << 6)
-            + encoded_operand
-    }
+#[derive(Copy, Clone, Hash, PartialEq, Eq, Debug)]
+#[repr(transparent)]
+pub(crate) struct AndOrPolarityOperandULE(u8);
 
-    #[inline]
-    fn validate_andor_polarity_operand(encoded: u8) -> Result<(), ZeroVecError> {
-        Operand::try_from(encoded & 0b0011_1111).map_err(|_| ZeroVecError::parse::<Self>())?;
+// Safety (based on the safety checklist on the ULE trait):
+//  1. AndOrPolarityOperandULE does not include any uninitialized or padding bytes
+//     (achieved by `#[repr(transparent)]` on a type that satisfies this invariant)
+/// 2. AndOrPolarityOperandULE is aligned to 1 byte
+//     (achieved by `#[repr(transparent)]` on a type that satisfies this invariant)
+//  3. The impl of validate_byte_slice() returns an error if any byte is not valid.
+//  4. The impl of validate_byte_slice() returns an error if there are extra bytes
+//     (impossible since it is of size 1 byte)
+//  5 The other ULE methods use the default impl.
+//  6. AndOrPolarityOperandULE byte equality is semantic equality.
+unsafe impl ULE for AndOrPolarityOperandULE {
+    fn validate_byte_slice(bytes: &[u8]) -> Result<(), ZeroVecError> {
+        for byte in bytes {
+            Operand::try_from(byte & 0b0011_1111).map_err(|_| ZeroVecError::parse::<Self>())?;
```

Comment:
Question: `Operand::try_from` is now the `impl ::core::convert::TryFrom<u8> for #name` from the ULE derive, correct?

Reply:
Yep!